File: scala/src/main/scala/com/example/cache/EhcacheUtils.scala
package com.example.cache

import java.util.function.BiConsumer
import redis.clients.jedis.Jedis
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.ObjectOutputStream
import java.io.ObjectInputStream
import java.io.ByteArrayInputStream
import org.ehcache.Cache

object EhcacheUtils {
    def get[K, V](cache: Cache[K, V]) = (key: K) => {
        Option(cache.get(key))
    }: Option[V]
    
    def put[K, V](cache: Cache[K, V]) = (key: K, value: V) => { 
        cache.put(key, value)
    }: Unit
}

File: java/src/main/java/com/example/cache/EhcacheUtils.java
package com.example.cache;
import java.util.Optional;
import java.util.function.BiConsumer;
import java.util.function.Function;

import org.ehcache.Cache;

public class EhcacheUtils {
    // get does not distinguish "null" from empty
    public static <K, V> Function<K, Optional<V>> get(Cache<K, V> cache) {
        return key -> Optional.ofNullable(cache.get(key));
    }
    
    public static <K, V> BiConsumer<K, Optional<V>> put(Cache<K, V> cache) {
        return (key, value) -> {
            if (value.isPresent()) {
                cache.put(key, value.get());
            } else {
                cache.remove(key);
            }
        };
    }
}
